app/controllers/scim_rails/scim_users_controller.rb

```diff
@@ -1,26 +1,110 @@
 module ScimRails
-  class ScimUsersController < ApplicationController
+  class ScimUsersController < ScimRails::ApplicationController
     def index
-      @users = User.last
-      scim_response(@users)
+      if params[:filter].present?
+        query = ScimRails::ScimQueryParser.new(params[:filter])
+
+        users = @company
+          .public_send(ScimRails.config.scim_users_scope)
+          .where(
+            "#{ScimRails.config.scim_users_model.connection.quote_column_name(query.attribute)} #{query.operator} ?",
+            query.parameter
+          )
+          .order(ScimRails.config.scim_users_list_order)
+
+        counts = ScimCount.new(
+          start_index: params[:startIndex],
+          limit: params[:count],
+          total: users.count
+        )
+        json_scim_response(object: users, counts: counts)
+      else
+        users = @company
+          .public_send(ScimRails.config.scim_users_scope)
+          .order(ScimRails.config.scim_users_list_order)
+
+        counts = ScimCount.new(
+          start_index: params[:startIndex],
+          limit: params[:count],
+          total: users.count
+        )
+        json_scim_response(object: users, counts: counts)
+      end
     end
 
     def create
+      user = @company.public_send(ScimRails.config.scim_users_scope).create!(permitted_user_params)
+      update_status(user) unless params[:active].nil?
+      json_scim_response(object: user, status: :created)
     end
 
     def show
+      user = @company.public_send(ScimRails.config.scim_users_scope).find(params[:id])
+      json_scim_response(object: user)
     end
 
-    def update
+    def put_update
+      user = @company.public_send(ScimRails.config.scim_users_scope).find(params[:id])
+      update_status(user) unless params[:active].nil?
+      user.update!(permitted_user_params)
```

Comment:
I actually thought a lot about this. And I don't know what the best solution is.

So for our app, we do not allow changes once someone is archived. Okta will however send records that need to be created and immediately archived (new folks not yet active). So if someone who has not yet started but needs their department updated then the update will fail on our end because we cannot update archived records. The only way to avoid this would be to allow updates to archived records or redefine what it means to be active/inactive.

For the same scenario where Okta sends a user to be created then immediately archived, when that user becomes active they will send a PUT request to change the status. If we try to update potentially different attributes on an archived user _first_ then the update will fail and the user will remain archived. But, if we change the status first _then_ update attributes then the request will be successful. That is the reason I went with archive/unarchive first.

For Okta to deprovision, they will send a PATCH request so we do not need to worry about changing other attributes.

Reply:
That makes sense. We care more about updating and re-provision a deprovisioned user than we do updating a user before they get deprovisioned. I'm good as-is 👍